File: src/com/cattailsw/nanidroid/install/NarRelativePathPolicy.java
package com.cattailsw.nanidroid.install;

import java.nio.charset.Charset;
import java.text.Normalizer;
import java.util.Locale;

/** Shared pure relative-path rules for NAR entries and live ghost trees. */
final class NarRelativePathPolicy {
    static final int MAX_DEPTH = 32;
    static final int MAX_PATH_BYTES = 1024;
    static final int MAX_COMPONENT_BYTES = 255;
    private static final Charset UTF_8 = Charset.forName("UTF-8");

    private NarRelativePathPolicy() {}

    static Result normalize(String raw) {
        if (raw == null
                || raw.length() == 0
                || raw.startsWith("/")
                || raw.indexOf('\\') >= 0
                || !validUnicode(raw)) {
            return Result.failure(Error.INVALID_PATH);
        }
        String[] components = raw.split("/", -1);
        if (components.length > MAX_DEPTH) {
            return Result.failure(Error.PATH_DEPTH_LIMIT);
        }
        StringBuilder normalized = new StringBuilder();
        try {
            for (String component : components) {
                if (component.length() == 0
                        || ".".equals(component)
                        || "..".equals(component)
                        || component.indexOf(':') >= 0
                        || containsControl(component)) {
                    return Result.failure(Error.INVALID_PATH);
                }
                String nfc = Normalizer.normalize(
                        component, Normalizer.Form.NFC);
                if (nfc.getBytes(UTF_8).length
                        > MAX_COMPONENT_BYTES) {
                    return Result.failure(
                            Error.COMPONENT_LENGTH_LIMIT);
                }
                if (normalized.length() > 0) {
                    normalized.append('/');
                }
                normalized.append(nfc);
            }
            String path = normalized.toString();
            if (path.getBytes(UTF_8).length > MAX_PATH_BYTES) {
                return Result.failure(Error.PATH_LENGTH_LIMIT);
            }
            return Result.success(
                    path, collisionKey(path));
        } catch (RuntimeException error) {
            return Result.failure(Error.INVALID_PATH);
        }
    }

    static String collisionKey(String value) {
        String nfc = Normalizer.normalize(
                value, Normalizer.Form.NFC);
        return Normalizer.normalize(
                nfc.toLowerCase(Locale.US),
                Normalizer.Form.NFC);
    }

    private static boolean containsControl(String value) {
        for (int index = 0; index < value.length(); index++) {
            if (Character.isISOControl(value.charAt(index))) {
                return true;
            }
        }
        return false;
    }

    private static boolean validUnicode(String value) {
        for (int index = 0; index < value.length(); index++) {
            char current = value.charAt(index);
            if (Character.isHighSurrogate(current)) {
                if (++index >= value.length()
                        || !Character.isLowSurrogate(
                                value.charAt(index))) {
                    return false;
                }
            } else if (Character.isLowSurrogate(current)) {
                return false;
            }
        }
        return true;
    }

    enum Error {
        INVALID_PATH,
        PATH_DEPTH_LIMIT,
        PATH_LENGTH_LIMIT,
        COMPONENT_LENGTH_LIMIT
    }

    static final class Result {
        private final String normalized;
        private final String key;
        private final Error error;

        private Result(
                String normalized, String key, Error error) {
            this.normalized = normalized;
            this.key = key;
            this.error = error;
        }

        private static Result success(
                String normalized, String key) {
            return new Result(normalized, key, null);
        }

        private static Result failure(Error error) {
            return new Result(null, null, error);
        }

        boolean isSuccess() { return error == null; }
        String getNormalized() { return normalized; }
        String getKey() { return key; }
        Error getError() { return error; }
    }
}

File: src/com/cattailsw/nanidroid/install/NarRelativePathPolicy.kt
package com.cattailsw.nanidroid.install

import java.nio.charset.StandardCharsets
import java.text.Normalizer
import java.util.Locale

/** Shared pure relative-path rules for NAR entries and live ghost trees. */
/**
 * This is internal to the modern app module; the frozen Java build uses its
 * matching source overlay under legacy/src instead.
 */
internal object NarRelativePathPolicy {
    const val MAX_DEPTH = 32
    const val MAX_PATH_BYTES = 1024
    const val MAX_COMPONENT_BYTES = 255

    @JvmStatic
    fun normalize(raw: String?): Result {
        if (raw.isNullOrEmpty() || raw.startsWith('/') || '\\' in raw || !validUnicode(raw)) {
            return Result.failure(Error.INVALID_PATH)
        }
        // Kotlin's split limit must be non-negative. Its default keeps the
        // trailing empty component too, matching Java's split("/", -1).
        val components = raw.split('/')
        if (components.size > MAX_DEPTH) return Result.failure(Error.PATH_DEPTH_LIMIT)
        return try {
            val normalized = StringBuilder()
            for (component in components) {
                if (component.isEmpty() || component == "." || component == ".." || ':' in component || containsControl(component)) return Result.failure(Error.INVALID_PATH)
                val nfc = Normalizer.normalize(component, Normalizer.Form.NFC)
                if (nfc.toByteArray(StandardCharsets.UTF_8).size > MAX_COMPONENT_BYTES) return Result.failure(Error.COMPONENT_LENGTH_LIMIT)
                if (normalized.isNotEmpty()) normalized.append('/')
                normalized.append(nfc)
            }
            val path = normalized.toString()
            if (path.toByteArray(StandardCharsets.UTF_8).size > MAX_PATH_BYTES) Result.failure(Error.PATH_LENGTH_LIMIT) else Result.success(path, collisionKey(path))
        } catch (_: RuntimeException) { Result.failure(Error.INVALID_PATH) }
    }

    @JvmStatic
    fun collisionKey(value: String): String = Normalizer.normalize(Normalizer.normalize(value, Normalizer.Form.NFC).lowercase(Locale.US), Normalizer.Form.NFC)
    private fun containsControl(value: String) = value.any { it.isISOControl() }
    private fun validUnicode(value: String): Boolean { var index = 0; while (index < value.length) { val current = value[index]; if (current.isHighSurrogate()) { if (++index >= value.length || !value[index].isLowSurrogate()) return false } else if (current.isLowSurrogate()) return false; index++ }; return true }
    enum class Error { INVALID_PATH, PATH_DEPTH_LIMIT, PATH_LENGTH_LIMIT, COMPONENT_LENGTH_LIMIT }
    class Result private constructor(private val normalized: String?, private val key: String?, private val error: Error?) {
        fun isSuccess() = error == null
        fun getNormalized() = normalized
        fun getKey() = key
        fun getError() = error
        companion object { fun success(normalized: String, key: String) = Result(normalized, key, null); fun failure(error: Error) = Result(null, null, error) }
    }
}
